doc/NAC/PnacSonicHLD.md

```diff
@@ -0,0 +1,374 @@
+- Enable / disable authentication globally
+- Enable / disable authentication on each port
+- Perform local authentication with secret in plain text 
+- Rate limit EAPOL packets to overcome/stop DDOS attack
+## 2.1 Target Deployment Use Cases
+First use case is Port based Network Access Control as described below.
+![PNAC Use Case](images/PNAC_Use_Case.png)
+Above diagram authenticator block/unblocks all traffic in ingress direction based on authentication status of the port. Traffic is allowed only for Authenticated users.
+# 3 Design
+## 3.1 Options to Block or allows port traffic based on Authentication status of the port
+As part of the design there are various design options were considered to effectively allow/disallow user traffic based on authentication status of the port connected to the end user device. Following paragraph list all these option with pros and cons and provide which option is best suited for the design.
+### 3.1.1 Options-1 Allowing or Blocking user traffic by using RSTP STP states 
+In this design option, if the NAC-enabled bridge-port is not authenticated then Port state is configured in “Blocked State” by invoking SAI API. After bridge-port get authenticated then the port state is changed to “Forwarding State”.
+
+***Pros:***
+PNAC implementation reuses existing SAI APIs for setting port state for controlling authorization status on a port.
+***Cons:***
+Currently SONiC distribution does not support STP/RSTP protocol.
+Port state should be set taking into account both NAC status on a port and RSTP/MSTP status. 
+In future, when RSTP/STP code gets implemented the port state need to be controlled using combination of NAC and RSTP/STP protocol. This is complicates the solution.
+### 3.1.2 Option-2: Allowing or Blocking user traffic by configuring L2-ACL filters 
+In this design option, software shall use two ACL filters,
+
+L2-ACL filter to allow EAPOL traffic
+L2-ACL filter to block the data traffic
+
+When a bridge port is in unauthorized state following filters shall be configured (index indicates priority of the filter).
+
+L2-ACL filter to allow EAPOL traffic
+L2-ACL filter to block the data traffic
+
+When a bridge port is in authorised state filters then following filter is configured.
+
+L2-ACL filter to allow EAPOL traffic
+
+
+***Pros:***
+Allowing and blocking traffic is achieved by L2-ACL filter using existing SAI APIs. Hence, no change in SAI API is required to achieve the functionality.  
+***Cons:***
+Total number of L2-ACL filters resources are limited for most of forwarding ASIC.
+Hence consuming 4 filters per interface will consume large number entries L2-ACL filter table, if NAC is enabled for interfaces. This will reduce useful number entries in L2-ACL filter table.
+### 3.1.3 Option-3: Allowing or Blocking user traffic by configuring learn mode
```

Comment:
Any unauthorized traffic that gets trapped to CPU can get VLAN flooded in the software. Need to address the same.

Reply:
In current implementation feature is dropping the traffic and not forwarding to CPU. For MAC based implementation we are planning to rate control traffic hitting CPU and this will be updated in the design document.